ADR 012: Multi-Version Support (#1274)
informalsystems/tendermint-rs: docs/architecture/adr-012-multi-version.md

}
```

The mode parameter is used to dynamically select the encoding used by a client.
By default, the clients use the latest protocol (designated by the associated
const `CompatMode::LATEST`), but it's possible to specify
the compatibility mode at construction using the newly introduced configuration
API. In the HTTP client, it's also possible to switch the mode for a client
that has been already connected; this is useful for dynamic version discovery
using the `status` endpoint.

The `Client` trait is extended to support methods new to CometBFT 0.37,
which are emulated using older protocol endpoints when `V0_34` compatibility
mode is selected. There is no need to deprecate any methods. All data types
in the public client API are the domain types of `tendermint`, so the
difference in encoding is confined to the crate internals.

### tendermint-abci

This crate is not actively supported and we should only make minimal effort
to make it work with the proposed changes in a single chosen version of the
protocol.

## Open questions

* Should we migrate tendermint-abci to 0.37, freeze it at 0.34 and declare
  end of life, or commit to the same level of multi-version support as in

>   other crates?

Reviewer: As discussed during the sync review, we've decided to migrate it to 0.37 in order to give its eventual customers a chance to update their apps for 0.37. We will eventually deprecate `tendermint-abci` in favor of [`tower-abci`](https://github.com/penumbra-zone/tower-abci) once the latter adds support for 0.37.

Let's update the ADR with this answer?

Author: OK. I'll add a negative consequence then: the existing `tendermint-abci` users who wish to remain on 0.34 will be effectively stranded on the previous release (0.29.x currently) of the crate.

Reviewer: Good point!

the CometBFT Core protocols and expect all chain and relayer operators to
upgrade their software in sync. But now there is a number of established chains
using CometBFT 0.34 in production, and new chains using future versions of the
protocol are expected to be deployed into the shared Cosmos ecosystem
so that relaying is possible between the old and the new chains.

The approach to protocol evolution and versioning in CometBFT has not been
very rigorous to date. As a result, the 0.37 release has a number of breaking
changes in the RPC protocol format. Though the changes seem beneficial
in the long term, there is only an ad-hoc way to discover which version is used
by the node and select the appropriate encoding. It appears to be too
late in the development process to introduce significant changes to mitigate
this in the 0.37 release timeframe. A requirement to specify the
protocol version in chain configuration would complicate the deployment
experience for chain and relayer operators and should be avoided.

Another motivation for multi-version support is versioning of tendermint-rs
itself: previously, the semver compatibility of tendermint-rs releases was

> tied with corresponding CometBFT Core version sequences in a none too obvious

Reviewer 2: ```suggestion
tied with corresponding CometBFT version sequences in a none too obvious
```

> way. It's desirable to decouple tendermint-rs versioning from that of CometBFT
> Core, making it solely a matter of Rust crate API evolution. In the long term,

Reviewer 2: ```suggestion
way. It's desirable to decouple tendermint-rs versioning from that of CometBFT,
making it solely a matter of Rust crate API evolution. In the long term,
```

a single version of the tendermint-rs libraries should be able to support all

> versions of CometBFT Core protocols that are relevant to the community.

Reviewer 2: ```suggestion
versions of CometBFT protocols that are relevant to the community.
```

The developers of CometBFT protocols should take versioning practices into use
for future revisions that support backward compatibility on the source code level.

## Decision

The tendermint-rs library API will be modified to provide support for

> multiple versions of CometBFT Core protocol, at this moment being 0.34 and

Reviewer 2: ```suggestion
multiple versions of CometBFT APIs, at this moment being 0.34 and
```

in the public client API are the domain types of `tendermint`, so the
difference in encoding is confined to the crate internals.

Protocol version discovery is not implemented in the library in a way that
would be invisible to the API user. To discover the Tendermint version, a client
needs to make a `status` request and process the version data from the response.
The format of the response is not divergent between the supported protocol
versions, which is why this should succeed regardless of the compatibility mode
initially selected. As the RPC client API represents individual endpoint requests,
it would be wrong to have the implementation perform a hidden RPC roundtrip to
discover the version when needed, and have the client exhibit interior
mutability that is otherwise not needed. In the future, a more formalized way
to discover the protocol version in use should be provided, so this ad-hoc
approach is expected to be eventually deprecated.

### tendermint-abci

This crate is not actively supported and we should only make the minimal effort
to update it to the 0.37 version of ABCI. No backward compatibility with 0.34

> or multi-versoin support is to be implemented; the consumers should be steered

Reviewer 2: ```suggestion
or multi-version support is to be implemented; the consumers should be steered
```
